### PR TITLE
Setup bastion host

### PR DESCRIPTION
**Pull request by PurpleBooth**
Currently this does not work as we have no way to get rescue system in place so we can install MicroOS

### CODE REVIEW

**Comment by mysticaltech** [comment]
Beautiful code right there; I am sure @phaer will be inspired by it. There must be a way to make this work, like setup with the public interface, then just take it down, all via userdata, or remote-exec (unless am not seeing things clearly).

File: bastion.tf [hunk 1]
```diff
@@ -0,0 +1,38 @@
+module "bastion" {
```
Comment (mysticaltech) [inline bastion.tf:1]: @PurpleBooth What is the role of this module?

File: modules/host/main.tf [hunk 2]
```diff
@@ -30,6 +30,16 @@ resource "hcloud_server" "server" {
   placement_group_id = var.placement_group_id
   user_data          = data.cloudinit_config.config.rendered
 
+  network {
+    ip         = var.private_ipv4
+    network_id = var.ipv4_subnet.network_id
+  }
+
+  public_net {
```
Comment (mysticaltech) [inline modules/host/main.tf:38]: Won't work that way, but not an issue. We can deploy a node and kill its public interface and access it through the private network.

Comment (mysticaltech) [reply modules/host/main.tf:38]: But the node needs to be deployed as explained at other places via bootstrapping through ignition/combusion. Very smart mechanism there, we created partition, and it auto-detected the scripts to execute before-boot, to enable or kill instances for that matter :)

**Comment by mysticaltech** [comment]
The beauty is that this project has solved a lot of similar scenarios, and has shared a few [links](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/tree/8634a1dc2fd5b05872cf2b35a5aae219ac691282) already to a past state in git, which should hopefully inspire you. The whole thing is there, the image, how to create the partitions, load the script, the reboot logic. With this, you just change a few lines to disable the interface and get your fully isolated agent nodes!

**Comment by mysticaltech** [comment]
Something important though, why? Do that? Why do we need to setup a bastion host. It's probably completely doable, but what benefit will we gain from it as a project, is I understand the need for agent nodes to be private and isolated, is the reason why? And in that case, why not just kill the public interface after everything is done with a remote exec? An attack during the terraform install is highly unlikely.

File: modules/host/main.tf (hunk 2 above)
Comment (mysticaltech) [reply modules/host/main.tf:38]: Interesting article started in the private ip only agent nodes task https://www.hetzner.com/news/06-22-flexibles-networking/. 
ksnip_20220802-164508

**Comment by mysticaltech** [comment]
@PurpleBooth As shared here, if possible, we can do private agents more simply, with bare Kubernetes, without having an additional bastion host. See my comment here https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/241#issuecomment-1203991195. 

The simpler, the better! But definitely, I would love to hear your arguments if you don't agree.

File: modules/host/versions.tf [hunk 3]
```diff
@@ -2,7 +2,7 @@ terraform {
   required_providers {
     hcloud = {
       source  = "hetznercloud/hcloud"
-      version = ">= 1.33.2"
+      version = ">= 1.34.2"
```
Comment (4ND3R50N) [inline modules/host/versions.tf:5]: Hey 👍 There are some bugs in this version related to private network only server!

I highly recommend updating the version to the latest one.

Please have a look here:
https://github.com/hetznercloud/terraform-provider-hcloud/releases/tag/v1.35.1

Comment (mysticaltech) [reply modules/host/versions.tf:5]: Thanks, @4ND3R50N; please don't hesitate to PR so we'll bump that version globally. Otherwise, I'll do it ASAP.

**Comment by mysticaltech** [comment]
@PurpleBooth While merging the latest changes, I noticed that you were probably assuming that `rebootmgr` takes care of all reboots including dealing with `kured`. This is not the case, in order to disable `kured`, you have to add a `REBOOT_METHOD=none` to the transactional_update.conf file, located somewhere in /etc.

More on this here https://en.opensuse.org/Kubic:Update_and_Reboot#Reboot_Strategy_Options and https://github.com/openSUSE/transactional-update/blob/master/etc/transactional-update.conf.

**Comment by mysticaltech** [comment]
Now I understand better what this is, thanks to your topology description. It also made me realize that node upgrades, k3s upgrades, and container fetching would disappear, which is not really ideal. Or am I mistaken?

**Comment by mysticaltech** [comment]
Ah, they do have a gateway that the private subnet can route traffic to in all likelihood. So the nodes would not be cut out from OUT traffic (but needs to be tested, of course).

**Comment by mysticaltech** [comment]
@PurpleBooth Interesting article https://docs.hetzner.com/cloud/servers/getting-started/connecting-via-private-ip/

**Comment by mysticaltech** [comment]
I believe the next step here is to use snapshots made with "normal" nodes. So yes, the bastion setup will take longer, but that's a small initial price to pay for the added sec.

**Comment by mysticaltech** [comment]
Just stumbled on this, seems interesting, but probably not needed (at least for now) https://github.com/inlets/inlets-pro

**Comment by mysticaltech** [comment]
@PurpleBooth I will close this PR for now, because it has stalled and the base is significantly outdated. But the branch will remain, so it can be picked up again later without issue.

**Pull request closed without merging**